### DISCUSSION TITLE
I still don't understand why we change where to put endpoint registration from UseRouting to UseEndpoint.

### DISCUSSION THREAD

**Discussion by aaronamm** [dotnet/aspnetcore#28157] [Q&A] [answered]
Hello everyone

During learning the difference between EndpointRoutingMiddleware and EndpointMiddleware, I still don't understand why we change where to put endpoint registration from UseRouting to UseEndpoint.

My curiosity has started from reading this blogpost [Understanding ASP.NET Core Endpoint Routing
](https://aregcode.com/blog/2019/dotnetcore-understanding-aspnet-endpoint-routing/).

From v3.0.0-preview4-19216-03 to v3.0.0-preview3-19153-02, we changed endpoint registration from UseRouting to UseEndpoint and here is the PR that changes endpoint registration.
https://github.com/dotnet/aspnetcore/pull/8765

The author of this PR described the purpose but I still don't understand it.
I understand that registration e.g. `MapGet` expose some execution logics and it is good to put it where we execute it.

Here is the flow that I understand, please correct me if it is incorrect.
- Application start
- Create endpoint list from registration defined in UseEndpoint(EndpointMiddleware)
- Start processing a web request
- UseRoute(EndpointRoutingMiddleware) selects an endpoint that match a request.
- UseEndpoint(EndpointMiddleware) executes an endpoint that turn out executing controller/action method and get a response back.

This makes me feel like the code jumps from EndpointMiddleware to EndpointRoutingMiddleware
From my understanding EndpointRoutingMiddleware should be the place to prepare and select a match route.

Therefore, could you please explain me more about this change?

Thank you so much.

**Comment by Tratcher** [comment, marked as answer]
"This is vital because it puts your code visually at the point of the pipeline where it executes."

The goal was to make the linear flow of startup intuitive, especially when you add middleware.

Consider this example:
```
            app.UseStaticFiles();

            app.UseRouting();

            app.UseAuthentication();

            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllerRoute(
                    name: "default",
                    pattern: "{controller=Home}/{action=Index}/{id?}");
                endpoints.MapRazorPages();
            });
```

UseEndpoints is where routes are specified to help indicate that middleware like UseAuthentication and UseAuthorization will run before the endpoints are executed.

> This makes me feel like the code jumps from EndpointMiddleware to EndpointRoutingMiddleware
> From my understanding EndpointRoutingMiddleware should be the place to prepare and select a match route.

I don't quite understand the concern here. EndpointRoutingMiddleware selects a route/endpoint and then EndpointMiddleware executes it. The pause in between lets other middleware see what endpoint was selected and react if they need to.

**Reply by aaronamm** [reply]
@Tratcher Thank you for answering my question.
When you gave code example with `UseEndpoints` that passes `endpoint` as a parameter,  this makes me can visualize the linear flow of startup. I can think the endpoint is executed after  `app.UseAuthentication()` and  `app.UseAuthorization()` middleware.

However, for the code example in https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/samples/MvcSandbox/Startup.cs#L40, we use `builder` as  a lambda parameter that makes me cannot visualize the linear flow of startup.
This makes me feel like why we put routing setup in `UseEndpoint` instead of inside  `UseRouting`.

Thank you.

**Reply by Tratcher** [reply]
That sandbox scenario is the same, I just used a different name for the builder variable. The actual type there is called `IEndpointRouteBuilder`.

**Reply by aaronamm** [reply]
@Tratcher Now it's clear to me, thanks.
